### ISSUE TITLE
[WICKET-7131] Improve accessibility and screen reader support for auto-complete component

### ISSUE THREAD

**Issue by jira-importer** [jira-importer/jira-to-github-issues#39]
<p>There are several accessibility issues with the auto-complete component that should be addressed to improve support for screen readers. <br/>
See <a href="https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-both/" class="external-link" target="_blank" rel="nofollow noopener">https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-both/</a> for examples and more details.</p>

<p>To summarize, the following improvements are needed:</p>
<ul>
	<li>The up and down arrow keys in the auto-complete dropdown should not move the cursor to the start or end of the text field</li>
	<li>aria-expanded should be set on the input element to indicate whether the dropdown is expanded or collapsed</li>
	<li>aria-selected should be set on the element that is currently selected in the dropdown</li>
	<li>aria-activedescendant attribute should be used on the input field, pointing to the currently selected item in the dropdown. This allows screen readers to track the active selection in the list.</li>
	<li>aria-live should not be set on the dropdown as it will cause a screen reader to read out everything in the dropdown, which disrupts the navigaition workflow</li>
</ul>

<p>Some of these issues were addressed in <a href="https://issues.apache.org/jira/browse/WICKET-7120" title="Improve Accessibility of AutoCompleteBehavior" class="issue-link" data-issue-key="WICKET-7120"><del>WICKET-7120</del></a>. But the current implementation relies on the dropdown being an unordered list, which may not work for all situations. And that issue where only merged to the main branch, which we cannot use due to Jakarta EE namespace changes in Wicket 10.</p>

<p>I will provide a pull request with these changes</p>

---
<details><summary><i>Originally reported by <a title="Erik Strid" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikstrid">erikstrid</a>, imported from: <a href="https://issues.apache.org/jira/browse/WICKET-7131" target="_blank">Improve accessibility and screen reader support for auto-complete component</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Martin Tzvetanov Grigorov" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mgrigorov">mgrigorov</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2024-11-19T09:13:56+00:00
<li><b>imported</b>: 2025-11-27
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>strido opened a new pull request, #1034:<br/>
URL: <a href="https://github.com/apache/wicket/pull/1034" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/1034</a></p>

<p>   Improved accessibility support with the following changes</p>

<ul class="alternate" type="square">
	<li>The up and down arrow keys in the auto-complete dropdown should not move the cursor to the start or end of the text field</li>
	<li>aria-expanded should be set on the input element to indicate whether the dropdown is expanded or collapsed</li>
	<li>aria-selected should be set on the element that is currently selected in the dropdown</li>
	<li>aria-activedescendant attribute should be used on the input field, pointing to the currently selected item in the dropdown. This allows screen readers to track the active selection in the list.</li>
	<li>aria-live should not be set on the dropdown as it will cause a screen reader to read out everything in the dropdown, which disrupts the navigaition workflow</li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>martin-g commented on PR #1034:<br/>
URL: <a href="https://github.com/apache/wicket/pull/1034#issuecomment-2484918266" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/1034#issuecomment-2484918266</a></p>

<p>   ```<br/>
      ../../wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/autocomplete/wicket-autocomplete.js<br/>
   <span class="error">&#91;INFO&#93;</span>     472 |var input = Wicket.$(ajaxAttributes.c);<br/>
   <span class="error">&#91;INFO&#93;</span>      ^ 'input' is already defined.<br/>
   <span class="error">&#91;INFO&#93;</span>     626 |    element.firstChild.role = "listbox"<br/>
   <span class="error">&#91;INFO&#93;</span> ^ Missing semicolon.<br/>
   <span class="error">&#91;INFO&#93;</span>     677 |    node.setAttribute("aria-posinset", i + 1)<br/>
   <span class="error">&#91;INFO&#93;</span>       ^ Missing semicolon.<br/>
   <span class="error">&#91;INFO&#93;</span>     678 |    node.setAttribute("aria-setsize", elementCount)<br/>
   <span class="error">&#91;INFO&#93;</span>     ^ Missing semicolon.<br/>
   ```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>strido commented on PR #1034:<br/>
URL: <a href="https://github.com/apache/wicket/pull/1034#issuecomment-2484979717" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/1034#issuecomment-2484979717</a></p>

<p>   The code style issues should be fixed now</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>martin-g merged PR #1034:<br/>
URL: <a href="https://github.com/apache/wicket/pull/1034" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/1034</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jira-bot">jira-bot</a>:</i>
<p>Commit b25c8311b95e46b46f8b23134b68adc3a4f2e543 in wicket's branch refs/heads/wicket-9.x from Erik Strid<br/>
[ <a href="https://gitbox.apache.org/repos/asf?p=wicket.git;h=b25c8311b9" class="external-link" target="_blank" rel="nofollow noopener">https://gitbox.apache.org/repos/asf?p=wicket.git;h=b25c8311b9</a> ]</p>

<p><a href="https://issues.apache.org/jira/browse/WICKET-7131" title="Improve accessibility and screen reader support for auto-complete component" class="issue-link" data-issue-key="WICKET-7131"><del>WICKET-7131</del></a> Improved accessibility and screen reader support (#1034)</p>

<ul>
	<li>Improved accessability and screen reader support</li>
</ul>


<ul>
	<li>Ensure that the input field exists before trying to change its attributes</li>
</ul>


<ul>
	<li>Fix code style issues</li>
</ul>


<p>---------</p>

<p>Co-authored-by: Erik Strid <erik.strid@idainfront.se></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jira-bot">jira-bot</a>:</i>
<p>Commit 219d45ccb3f9a9069295eb53da9d9a7141fd225c in wicket's branch refs/heads/master from Erik Strid<br/>
[ <a href="https://gitbox.apache.org/repos/asf?p=wicket.git;h=219d45ccb3" class="external-link" target="_blank" rel="nofollow noopener">https://gitbox.apache.org/repos/asf?p=wicket.git;h=219d45ccb3</a> ]</p>

<p><a href="https://issues.apache.org/jira/browse/WICKET-7131" title="Improve accessibility and screen reader support for auto-complete component" class="issue-link" data-issue-key="WICKET-7131"><del>WICKET-7131</del></a> Improved accessibility and screen reader support (#1034)</p>

<ul>
	<li>Improved accessability and screen reader support</li>
</ul>


<ul>
	<li>Ensure that the input field exists before trying to change its attributes</li>
</ul>


<ul>
	<li>Fix code style issues</li>
</ul>


<p>---------</p>

<p>Co-authored-by: Erik Strid <erik.strid@idainfront.se><br/>
(cherry picked from commit b25c8311b95e46b46f8b23134b68adc3a4f2e543)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mgrigorov">mgrigorov</a>:</i>
<p>The improvement is applied to 9.x and master branches.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jira-bot">jira-bot</a>:</i>
<p>Commit dc3b7c747cccb343c8951677507563dec9091462 in wicket's branch refs/heads/wicket-10.x from Erik Strid<br/>
[ <a href="https://gitbox.apache.org/repos/asf?p=wicket.git;h=dc3b7c747c" class="external-link" target="_blank" rel="nofollow noopener">https://gitbox.apache.org/repos/asf?p=wicket.git;h=dc3b7c747c</a> ]</p>

<p><a href="https://issues.apache.org/jira/browse/WICKET-7131" title="Improve accessibility and screen reader support for auto-complete component" class="issue-link" data-issue-key="WICKET-7131"><del>WICKET-7131</del></a> Improved accessibility and screen reader support (#1034)</p>

<ul>
	<li>Improved accessability and screen reader support</li>
</ul>


<ul>
	<li>Ensure that the input field exists before trying to change its attributes</li>
</ul>


<ul>
	<li>Fix code style issues</li>
</ul>


<p>---------</p>

<p>Co-authored-by: Erik Strid <erik.strid@idainfront.se><br/>
(cherry picked from commit b25c8311b95e46b46f8b23134b68adc3a4f2e543)</p>
